### ISSUE TITLE
REQ: Exit-save-dialog behavior with auto-backup/snapshot: Default or untangle as extra option

### ISSUE THREAD

**Issue by kxrob** [notepad-plus-plus/notepad-plus-plus#9258]
np++ since some time has the snapshot & periodic backup mode - which even is the default (`isSnapshotMode="yes"`)

The auto-backup is generally a good feature to not loose content - should remain the default.

Entangled however is the behavior that there is no question for file save upon np++ exit: _"If this feature is enabled (enabled by default), user won't be asked to save unsaved file as he quits Notepad++, and on startup Notepad++ restores the unsaved file and unsaved untitled document of last session."_

This behavior is quite UNACCEPTABLE and dangerous in most cases. E.g. when using np++ with git, which is even suggested by its installer one usually forgets to save during git rebase -i  , message edits etc. - and false things happen. The same for most technical file edits when subsequent operations, builds, commits, file syncs, etc. are dependent on the edits. This constantly causes trouble.
And switching off this entangled default option, one looses the periodic backup and snapshot behavior completely.

No other editor, IDE, Libre app, Word, Excel, or whatever I know of does silently close with unsaved changes !? At least not by default.

See also the comments and their upvotes on this stackoverflow answer here:
[Notepad ++ doesn't save document on exit?](https://stackoverflow.com/questions/24447786/notepad-doesnt-save-document-on-exit/24448488#24448488)

=> So the exit-save-dialog should be used either always (like in all other decent editor-like progs),  or should at least be offered as extra option when periodic backup is on and in this case be ON by default (because silent exit is dangerous and surprising.)
When the dialog is answered with NO it should act according the snapshot/backup setting (thus potentially preserve the unsaved changes for next startup) and propagate to the question for the next changed file (without Cancel). Thus with snapshot/backup ON the user needs to close a tab explicitly (or "Reload from disk") to get rid of unsaved changes completely in the end.

**Comment by sasumner** [comment]
Why not just turn off the ability to exit with unsaved documents, and turn on the backup-upon-save; and then save relatively often?

**Comment by sasumner** [comment]
>No other editor, IDE, Libre app, Word, Excel, or whatever I know of does silently close with unsaved changes !? 

I think this was a feature that users liked, and was a product-differentiator at one time ("Use Notepad++, it can do this cool thing!").
Personally, I've never liked that, and as a user I've never had that turned on.
And perhaps, as you point out, there are some definite downsides to that kind of action.

**Comment by kxrob** [comment]
> and then save relatively often?

Which means sacrifice auto periodic backup - and always have extra pressure in mind during longer edit sessions.
The computer should do that.

> Why not just turn off the ability to exit with unsaved documents, 

Exit with unsaved could still be possible (there are reasons in (fewer) cases) - but only after explicit confirmation  ("No" button to save question), not silently.

> that users liked

.. and (only) then I may also like that option.

> and turn on the backup-upon-save; 

Which is a different game: preserve previous verion(s) of the file.

**Comment by sasumner** [comment]
>...and always have extra pressure in mind during longer edit sessions.
>The computer should do that.

I seem to get into these kinds of discussions a lot with people. :-)
About what _should_ go on, versus the reality of protecting your data in the best possible manner.
But you are absolutely right; I can't argue with "the computer should do that".

Regarding your request, I'm not the absolute authority, but I can't really see it being taken on.
I wouldn't touch it, as too disruptive to the entrenched behavior.
But good luck to you; maybe you'll consider self-implementing it and contributing it?

**Comment by sasumner** [comment]
I considered this further...

WIth a scripting plugin, you could set up a script that iterates through all of the tabs in N++ and saves the ones that are named files on disk (or you could prompt to save/discard same) and then exits Notepad++.  Thus, the script would ignore files like "new 2", "new 5", etc.
The idea here is that you would enable the option in N++ preferences that allows you to exit while keeping unsaved files.
Then you run the script when you want to exit N++ (easy to set up as toolbar button or keycombo).
The files that need to be saved get saved, N++ remembers the data for any "new X" files, you get the benefits of periodic backup on all files (and of course also have the option to do backup-on-save additionally) and it seems like the world then becomes a happy place.
A limitation might be that it takes a bit of time to run if you have a lot of tabs (it's the switching of the active document that is time-intensive).  But if you're quitting anyway, just ignore it and start working with another app.

A kludge?  Well, maybe, but scripting plugins are a long-accepted method of "filling in the gaps" when you want Notepad++ to do something it doesn't quite do.

Anyway, just some food for thought.

**Comment by donho** [comment]
@kxrob 
> This behavior is quite UNACCEPTABLE and dangerous in most cases. E.g. when using np++ with git, which is even suggested by its installer one usually forgets to save during git rebase -i , message edits etc. - and false things happen.

I use also Notepad++ as git's editor. My backup feature is always ON and the described situation above has never happened to me. Could you provide the way to produce the issue you described?

**Comment by kxrob** [comment]
Note, this problem occurs when the "Enable session snapshot and periodic backup"  setting/checkbox in ON !
( There is also "Backup on save" , a different thing,  and "Remember current session..." )

* Edit the git issue 
* Close np++ (Alt-F4)  without having saved via Ctrl+S/Save  or Tab-close (which asks)
=> np++ does not ask for save , closes silently and the git issue runs off as if there was not edit. Same for any operation depending on a edited file being saved in its place (an not just in the np++ snapshot backup folder)

**Comment by sasumner** [comment]
@donho said:

>...and the described situation above has never happened to me

Because there is no PBKAC ?  ;-)

**Comment by sasumner** [comment]
@donho 

In seriousness, it might be nice to have a way for "new 2" files to persist across runs of Notepad++ as unsaved/untitled "red" documents, and yet still be prompted upon N++ exit to save/discard changes to any hard-named disk files.  What do you think?

**Comment by donho** [comment]
@kxrob 
My checkbox of "Enable session snapshot and periodic backup" setting is ON.
Then I modify a file, add it and commit it, Notepad++ is launched, type somethings and Alt-F4, the save dialog popup - so I cannot reproduce this issue at all. Could you provide your debug info please?

@sasumner 
> Because there is no PBKAC ? ;-)

Could be, I'm not perfect :) But if you can reproduce this issue as well, then I'm missing something absolutely. 

> In seriousness, it might be nice to have a way for "new 2" files to persist across runs of Notepad++ as unsaved/untitled "red" documents, and yet still be prompted upon N++ exit to save/discard changes to any hard-named disk files. What do you think?

I always prefer simple and stupid solution. User can turn off  "Enable session snapshot and periodic backup" feature if they need to save their document explicitly. The issue @kxrob has mentioned (if it exits) it need to be fixed, but not to change the current behaviour.

**Comment by sasumner** [comment]
@donho 

Sometimes our language barrier is the hardest part. :-)

I'm confused about your stated workflow:

>My checkbox of "Enable session snapshot and periodic backup" setting is ON.
Then I modify a file, add it and commit it, Notepad++ is launched, type somethings and Alt-F4, the save dialog popup...

Why would the save dialog pop up in this situation?
With _Enable session snapshot and periodic backup_ set to ON, a save dialog should not appear.

**_UNLESS_**, you already have Notepad++ running in another instance, and the one that git invokes is a _second_ instance (in which case _Enable session snapshot and periodic backup_ is set to OFF, but only for that _second_ instance, causing a Save dialog popup to happen).  Is your git configuration for this set to run N++ with the `-multiInst` parameter?

**Comment by donho** [comment]
@sasumner 
> UNLESS, you already have Notepad++ running in another instance, and the one that git invokes is a second instance (in which case Enable session snapshot and periodic backup is set to OFF, but only for that second instance, causing a Save dialog popup to happen).

It was the case on my last test.
However, I test it again WITHOUT any instance of Notepad++, modify a file, add it and commit. After typing somethings and close Notepad++ without saving - the save dialog popup.

>  Is your git configuration for this set to run N++ with the -multiInst parameter?

I guess so, because git launches always the 2nd Notepad++ instance. I installed (or updated) git by enabling notepad++.exe as the default editor WITHOUT modifying anything. It works perfectly. If you guys try to modifying git settings manually, add `-multiInst` will solve your problem.

**Comment by sasumner** [comment]
@donho I'm not actually trying the git setup, because I don't want to mess with mine.  I just have my git set to use notepad.exe for that simple stuff.

I think that @kxrob 's request isn't exclusively because of git interaction, that's just a good example to illustrate the request.

**Comment by donho** [comment]
As I said I prefer simple and stupid solution so I don't think we should add another option in an option. 
The real question to me is "Should we make `Enable session snapshot and periodic backup` ON by default"?
But the link in StackOverFlow has answered this question - if people don't want this behaviour, they can disable it.
Where as the git situation, usage of `-multiInst` is the solution. So no, this behaviour won't be changed.

**Comment by sasumner** [comment]
@kxrob So if you want the prompt-for-save-on-quit functionality PLUS the seven-second-backup functionality, which is what I think you've been saying all along, we just have to do a solution with a scripting plugin to make it happen.

**Comment by kxrob** [comment]
Yes, the question is to get the periodic backup - which is the (only) thing advertised by "_Enable session snapshot and periodic backup_" - AND being asked (by default) at exit. At least for established files.

So that things like in this upvoted comment at stackoverflow do not happen: 
"_at least they should've made it disabled by default! nearly messed up all my test databases thanks to this as a change in a sql script didn't get saved!_"

I wonder if any programmer or non-programmer ever would expect anything else. 
The only reasonable exception regarding skip of exit-save-dialog mentioned were new tabs not established as files so far - a trivial case differentiation:  
"_I often create new editors (tabs) and paste in some snippets. I do not want to create a new file for each of those editors, but I want to close Notepad++ without loosing all those temporary snippets._" 

The git example (thought w/o -nosession,  -multiinst) was just one of endless / majority of examples where subsequent actions depend on files - which is the normal thing with programming, admin and outside. Better forget git, because in that case its less import to have periodic backup and one can get away by exception with those parameters. 

The current (default) situation is particularly dangerous with multiple tabs (which pile up even more likely in that mode), because in np++ you don't see directly which tabs (except the active one) are not saved currently - unlike in SciTE itself where you see the "change-star" (*) directly for each tab.

I think there is no problem to go to a new default of ask-on-save for established files - because we'd go to a far less dangerous, more safe default. People could still re-enable the "strange" silent mode via an extra option - upon the first question - in case such people do exist. But I think an (extra) option for a complete silent exit without any question is not needed at all (--> simplicity). For one changed tab w established file its not much of an issue to answer the question with NO in rare cases. And for several files its anyway a hairy situation and mostly one would want to save a least part of the changed mess.

A plugin for such essential thing would confuse even more. I install/copy np++ als light weigth editor (alternative to SciTE) here and there and usually won't remember to get plugins up to prevent a dangerous data loss.

**Comment by sasumner** [comment]
>because in np++ you don't see directly which tabs (except the active one) are not saved currently

Turning on the new "alternate icons" feature aids this situation, as the saved/modified/locked tab icons are differentiated by shapes instead of colors (the colors being ineffective when a tab is not active).

>A plugin for such essential thing would confuse even more...

Sure, your choice.
I'm just trying to propose workarounds since it appears the request is heading toward being declined (at least I see no indication from the previous discussion of anything else).

**Comment by donho** [comment]
> Turning on the new "alternate icons" feature aids this situation, as the saved/modified/locked tab icons are differentiated by shapes instead of colors (the colors being ineffective when a tab is not active).

I'm not for adding stuffs & stuffs to complicate the feature.
The feedbacks I've got from a lot of users about this feature are positive.
